apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/ExternalPluginClassLoader.java

```diff
@@ -55,7 +58,27 @@ public ExternalPluginClassLoader(File pluginJar, ClassLoader agentClassLoader) t
             while (entries.hasMoreElements()) {
                 JarEntry jarEntry = entries.nextElement();
                 if (jarEntry.getName().endsWith(".class")) {
-                    tempClassNames.add(jarEntry.getName().replace('/', '.').substring(0, jarEntry.getName().length() - 6));
+                    String fqcn = jarEntry.getName().replace('/', '.').substring(0, jarEntry.getName().length() - 6);
+                    if (fqcn.startsWith("org.slf4j") || fqcn.startsWith("org.apache.logging")) {
+                        throw new IllegalStateException(String.format("Package \"%s\" is used within plugin %s. This is not allowed " +
+                                "because it is already used by the agent. For logging purposes, use the agent SDK logging facade - %s",
+                            fqcn.substring(0, fqcn.lastIndexOf('.')),
+                            pluginJar.getName(),
+                            LoggerFactory.class.getName()
+                        ));
+                    }
+                    for (String agentDependencyPackage : AgentInfo.getAgentDependencyPackages()) {
```

Comment:
Could we make the `ExternalPluginClassLoader` child-first and allow plugins to have a dependency on `HdrHistrogram`, for example?

Reply:
It will create further complications...
We create an `IndyPluginClassLoader` for each external plugin, which can load most of the plugin classes. If you leave it as is (i.e. include these libraries within the IndyPlugin CL classpath), then you may get `LinkageError` if the `ExternalPluginClassLoader ` also loads these classes (e.g. because the instrumentation class uses it), because it is in the plugin CL delegation hierarchy. If you exclude these classes from the ones loaded by the Indy Plugin CL, then by default it would delegate loading requests of them to the agent CL (because this is how the `DiscriminatingMultiParentClassLoader` works), which may lead to other `LinkageError`